File: src/main/java/org/eln2/sim/mna/component/Inductor.java
package org.eln2.sim.mna.component;

/*
NOTE: DO NOT IMPORT MINECRAFT CODE IN THIS CLASS
EXTEND IT INSTEAD IN THE org.eln.nbt DIRECTORY
 */

import org.eln2.sim.mna.SubSystem;
import org.eln2.sim.mna.misc.ISubSystemProcessI;
import org.eln2.sim.mna.state.CurrentState;
import org.eln2.sim.mna.state.State;

public class Inductor extends Bipole implements ISubSystemProcessI {//, INBTTReady {

    String name;

    private double l = 0;
    double ldt;

    private CurrentState currentState = new CurrentState();

    public Inductor(String name) {
        this.name = name;
    }

    public Inductor(String name, State aPin, State bPin) {
        super(aPin, bPin);
        this.name = name;
    }

    @Override
    public double getI() {
        return currentState.getState();
    }

    public double getL() {
        return l;
    }

    public void setL(double l) {
        this.l = l;
        dirty();
    }

    public double getE() {
        final double i = getI();
        return i * i * l / 2;
    }

    @Override
    public void applyTo(SubSystem s) {
        ldt = -l / s.getDt();

        s.addToA(getAPin(), currentState, 1);
        s.addToA(getBPin(), currentState, -1);
        s.addToA(currentState, getAPin(), 1);
        s.addToA(currentState, getBPin(), -1);
        s.addToA(currentState, currentState, ldt);
    }

    @Override
    public void simProcessI(SubSystem s) {
        s.addToI(currentState, ldt * currentState.getState());
    }

    @Override
    public void quitSubSystem() {
        getSubSystem().states.remove(getCurrentState());
        getSubSystem().removeProcess(this);
        super.quitSubSystem();
    }

    @Override
    public void setSubSystem(SubSystem s) {
        super.setSubSystem(s);
        s.addState(getCurrentState());
        s.addProcess(this);
    }

    public CurrentState getCurrentState() {
        return currentState;
    }

    /*
    @Override
    public void readFromNBT(NBTTagCompound nbt, String str) {
        str += name;
        currentState.state = (nbt.getDouble(str + "Istate"));
    }

    @Override
    public void writeToNBT(NBTTagCompound nbt, String str) {
        str += name;
        nbt.setDouble(str + "Istate", currentState.state);
    }

     */

    public void resetStates() {
        currentState.setState(0);
    }
}

File: src/main/java/org/eln2/sim/mna/component/Inductor.kt
package org.eln2.sim.mna.component

/*
NOTE: DO NOT IMPORT MINECRAFT CODE IN THIS CLASS
EXTEND IT INSTEAD IN THE org.eln.nbt DIRECTORY
 */


import org.eln2.sim.mna.SubSystem
import org.eln2.sim.mna.misc.ISubSystemProcessI
import org.eln2.sim.mna.state.CurrentState
import org.eln2.sim.mna.state.State

open class Inductor: Bipole, ISubSystemProcessI {

    constructor()
    constructor(aPin: State?, bPin: State?): super(aPin, bPin)

    open var currentState: CurrentState = CurrentState()

    open var l: Double = 0.0
        set(l) {
            field = l
            dirty()
        }

    override fun getI(): Double {
        return currentState.state
    }

    open fun getE(): Double {
        val i = getI()
        return i * i * l / 2
    }

    override fun applyTo(s: SubSystem) {
        val ldt = -l / s.dt
        s.addToA(aPin, currentState, 1.0)
        s.addToA(bPin, currentState, -1.0)
        s.addToA(currentState, aPin, 1.0)
        s.addToA(currentState, bPin, -1.0)
        s.addToA(currentState, currentState, ldt)
    }

    override fun simProcessI(s: SubSystem) {
        val ldt = -l / s.dt
        s.addToI(currentState, ldt * currentState.state)
    }

    override fun quitSubSystem(s: SubSystem) {
        s.removeState(currentState)
        s.processI.remove(this)
        super.quitSubSystem(s)
    }

    override var subSystem: SubSystem? = null
        set(s) {
            field = s
            s?.addState(currentState)
            s?.processI?.add(this)
        }

    open fun resetStates() {
        currentState.state = 0.0
    }
}

/*
    @Override
    public void readFromNBT(NBTTagCompound nbt, String str) {
        str += name;
        currentState.state = (nbt.getDouble(str + "Istate"));
    }

    @Override
    public void writeToNBT(NBTTagCompound nbt, String str) {
        str += name;
        nbt.setDouble(str + "Istate", currentState.state);
    }

     */